[Bug] [CLI] PSS report does not show properties with control details

coder12git opened the pull request:
## Explanation

<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->
Add `properties` field in pss report
## Related issue

<!--
Please link the GitHub issue this pull request resolves in the format of `Closes #1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->
Fixes #9784
## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## Documentation (required for features)

My PR contains new or altered behavior to Kyverno. 
- [ ] I have sent the draft PR to add or update [the documentation](https://github.com/kyverno/website) and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->

## What type of PR is this

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->
/kind bug
## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->
```
./cmd/cli/kubectl-kyverno/kubectl-kyverno apply tests/restricted.yaml -r tests/resources.yaml --policy-report

Applying 3 policy rule(s) to 1 resource(s)...
----------------------------------------------------------------------
POLICY REPORT:
----------------------------------------------------------------------
apiVersion: wgpolicyk8s.io/v1alpha2
kind: ClusterPolicyReport
metadata:
  creationTimestamp: null
  name: merged
results:
- message: 'Validation rule ''restricted'' failed. It violates PodSecurity "restricted:latest":
    (Forbidden reason: allowPrivilegeEscalation != false, field error list: [spec.containers[0].securityContext.allowPrivilegeEscalation:
    Required value])(Forbidden reason: unrestricted capabilities, field error list:
    [spec.containers[0].securityContext.capabilities.drop: Required value])(Forbidden
    reason: runAsNonRoot != true, field error list: [spec.containers[0].securityContext.runAsNonRoot:
    Required value])(Forbidden reason: seccompProfile, field error list: [spec.containers[0].securityContext.seccompProfile.type:
    Required value])'
  policy: psa
  properties:
    controls: allowPrivilegeEscalation,capabilities_restricted,runAsNonRoot,seccompProfile_restricted
    standard: restricted
    version: latest
  resources:
  - apiVersion: v1
    kind: Pod
    name: nginx
    namespace: default
  result: fail
  rule: restricted
  scored: true
  source: kyverno
  timestamp:
    nanos: 0
    seconds: 1708756718
summary:
  error: 0
  fail: 1
  pass: 0
  skip: 0
  warn: 0
```



## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

vishal-chdhry commented:
@coder12git lint is failing

coder12git commented:
> @coder12git lint and all conformance tests are failing

Ok, but why conformance tests are failing, i'm not able to find that

coder12git commented:
@vishal-chdhry i think now most of the tests are passed!

vishal-chdhry commented on cmd/cli/kubectl-kyverno/report/report.go:
```diff
@@ -54,6 +57,23 @@ func ComputePolicyReportResult(auditWarn bool, engineResponse engineapi.EngineRe
+			}
+		}
+		if len(controls) > 0 {
+			sort.Strings(controls)
```
any specific reason to do a sort operation here?

coder12git replied:
this part of code is present [here](https://github.com/kyverno/kyverno/blob/d737138ad51f71e5b7511ac53d01705fccc0bedc/pkg/utils/report/results.go#L114-L130), i just shifted that logic to this file

coder12git replied:
the main reason here is just make it more readable, by sorting it alphabetically like this -
 `controls: allowPrivilegeEscalation,capabilities_restricted,runAsNonRoot,seccompProfile_restricted`

vishal-chdhry reviewed (commented):
Can you please add a unit test for this too

coder12git commented:
> Can you please add a unit test for this too

Sure, I need to add that in [report_test.go](https://github.com/kyverno/kyverno/blob/main/cmd/cli/kubectl-kyverno/report/report_test.go), right?

JimBugwadia commented:
@coder12git - please add tests, so we can merge. Thanks!

coder12git commented:
> @coder12git - please add tests, so we can merge. Thanks!

Done, please check

vishal-chdhry commented:
@coder12git Tests failing, and I think there should be a test where the `properties` field is filled, as thats what this PR

coder12git commented:
> @coder12git Tests failing, and I think there should be a test where the `properties` field is filled, as thats what this PR adds

I have added it like this, is this the right way?

```
{
		name:           "fail",
		auditWarn:      false,
		engineResponse: engineapi.NewEngineResponse(unstructured.Unstructured{}, engineapi.NewKyvernoPolicy(policy), nil),
		ruleResponse:   *engineapi.RuleWarn("xxx", engineapi.Mutation, "test"),
		want: policyreportv1alpha2.PolicyReportResult{
			Source: "kyverno",
			Policy: "psa",
			Properties: map[string]string{
				"controls": strings.Join([]string{"allowPrivilegeEscalation", "capabilities_restricted", "runAsNonRoot", "seccompProfile_restricted"}, ","),
				"standard": "restricted",
				"version":  "xxx",
			},
			Rule:      "xxx",
			Result:    policyreportv1alpha2.StatusError,
			Resources: []corev1.ObjectReference{{}},
			Message:   "test",
			Scored:    true,
			Category:  "Pod Security Standards (Restricted)",
			Severity:  policyreportv1alpha2.SeverityMedium,
		},
	}
```

MariamFahmy98 commented:
There is an additional scenario where a resource is skipped due to an exception, which includes the `exception` field in the properties. Can you add it as well?

Here is how we add it to the policy reports: https://github.com/kyverno/kyverno/blob/main/pkg/utils/report/results.go#L109-L113.

coder12git commented:
> There is an additional scenario where a resource is skipped due to an exception, which includes the `exception` field in the properties. Can you add it as well?
> 
> Here is how we add it to the policy reports: https://github.com/kyverno/kyverno/blob/main/pkg/utils/report/results.go#L109-L113.

Sure, I will, can you please let me know that whether i have write the tests in a correct way or not?

coder12git commented:
> > There is an additional scenario where a resource is skipped due to an exception, which includes the `exception` field in the properties. Can you add it as well?
> > Here is how we add it to the policy reports: https://github.com/kyverno/kyverno/blob/main/pkg/utils/report/results.go#L109-L113.
> 
> Sure, I will, can you please let me know that whether i have wrote the tests in a correct way or not bcoz the tests are failing?

JimBugwadia commented:
I took a closer look. 

Rather that duplicate the logic in (EngineResponseToReportResults)[https://github.com/kyverno/kyverno/blob/d737138ad51f71e5b7511ac53d01705fccc0bedc/pkg/utils/report/results.go#L89] why don't we make it reusable, by extracting a `ToPolicyReportResult` method that works for all cases? This way we do not risk any future divergence, and avoid duplicated code.

```go
func ToPolicyReportResult(
	policyType engineapi.PolicyType,
	policyName string,
	ruleResult engineapi.RuleResponse,
	annotations map[string]string,
	resource *v1.ObjectReference) policyreportv1alpha2.PolicyReportResult {

	result := policyreportv1alpha2.PolicyReportResult{
		Source:  kyverno.ValueKyvernoApp,
		Policy:  policyName,
		Rule:    ruleResult.Name(),
		Message: ruleResult.Message(),
		Result:  toPolicyResult(ruleResult.Status()),
		Scored:  annotations[kyverno.AnnotationPolicyScored] != "false",
		Timestamp: metav1.Timestamp{
			Seconds: time.Now().Unix(),
		},
		Category: annotations[kyverno.AnnotationPolicyCategory],
		Severity: SeverityFromString(annotations[kyverno.AnnotationPolicySeverity]),
	}
	if result.Result == "fail" && !result.Scored {
		result.Result = "warn"
	}
	if resource != nil {
		result.Resources = []v1.ObjectReference{
			*resource,
		}
	}
	if ruleResult.Exception() != nil {
		result.Properties = map[string]string{
			"exception": ruleResult.Exception().Name,
		}
	}
	pss := ruleResult.PodSecurityChecks()
	if pss != nil && len(pss.Checks) > 0 {
		var controls []string
		for _, check := range pss.Checks {
			if !check.CheckResult.Allowed {
				controls = append(controls, check.ID)
			}
		}
		if len(controls) > 0 {
			sort.Strings(controls)
			result.Properties = map[string]string{
				"standard": string(pss.Level),
				"version":  pss.Version,
				"controls": strings.Join(controls, ","),
			}
		}
	}
	if policyType == engineapi.ValidatingAdmissionPolicyType {
		result.Source = "ValidatingAdmissionPolicy"
		if ruleResult.ValidatingAdmissionPolicyBinding() != nil {
			result.Properties = map[string]string{
				"binding": ruleResult.ValidatingAdmissionPolicyBinding().Name,
			}
		}
	}

	return result
}

```


```go
func EngineResponseToReportResults(response engineapi.EngineResponse) []policyreportv1alpha2.PolicyReportResult {
	pol := response.Policy()
	policyName, _ := cache.MetaNamespaceKeyFunc(pol.AsKyvernoPolicy())
	policyType := pol.GetType()
	annotations := pol.GetAnnotations()

	var results []policyreportv1alpha2.PolicyReportResult
	for _, ruleResult := range response.PolicyResponse.Rules {
		result := ToPolicyReportResult(policyType, policyName, ruleResult, annotations, nil)
		results = append(results, result)
	}

	return results
}

```

CLI code:

```go

func ComputePolicyReportResult(auditWarn bool, engineResponse engineapi.EngineResponse, ruleResponse engineapi.RuleResponse) policyreportv1alpha2.PolicyReportResult {
	policy := engineResponse.Policy()
	policyType := policy.GetType()
	policyName := cache.MetaObjectToName(policy.MetaObject()).String()
	resource := engineResponse.Resource
	resorceRef := &v1.ObjectReference{
		Kind:            resource.GetKind(),
		Name:            resource.GetName(),
		Namespace:       resource.GetNamespace(),
		UID:             resource.GetUID(),
		APIVersion:      resource.GetAPIVersion(),
		ResourceVersion: resource.GetResourceVersion(),
	}

	result := reportutils.ToPolicyReportResult(policyType, policyName, ruleResponse, policy.GetAnnotations(), resorceRef)
	if result.Result == policyreportv1alpha2.StatusFail {
		audit := engineResponse.GetValidationFailureAction().Audit()
		if audit && auditWarn {
			result.Result = policyreportv1alpha2.StatusWarn
		}
	}

	return result
}
```

JimBugwadia commented:
There is other logic in `EngineResponseToReportResults` for handling VAPs which also seemed to be missing in the CLI code.

JimBugwadia commented:
@coder12git - any thoughts or questions on my suggestions above?

coder12git commented:
> @coder12git - any thoughts or questions on my suggestions above?

I agree, we shouldn't duplicate the logic! I will implement it by today.

JimBugwadia commented on pkg/utils/report/results.go:
```diff
@@ -86,72 +87,74 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
-				},
+		if len(controls) > 0 {
+			sort.Strings(controls)
+			result.Properties = map[string]string{
```
This could potentially replace properties set on line 113. May be best to create an empty map for properties in the declaration on line 91, and then add properties to that for the exception and PSS controls.

coder12git replied:
@JimBugwadia i have updated that, please have a look

JimBugwadia commented on cmd/cli/kubectl-kyverno/report/report.go:
```diff
@@ -1,9 +1,9 @@
 
 import (
-	"github.com/kyverno/kyverno/api/kyverno"
+	// "github.com/kyverno/kyverno/api/kyverno"
```
Please remove commented lines...

JimBugwadia commented on cmd/cli/kubectl-kyverno/report/report.go:
```diff
@@ -1,9 +1,9 @@
+	// "github.com/kyverno/kyverno/api/kyverno"
 	policyreportv1alpha2 "github.com/kyverno/kyverno/api/policyreport/v1alpha2"
-	"github.com/kyverno/kyverno/cmd/cli/kubectl-kyverno/policy/annotations"
+	// "github.com/kyverno/kyverno/cmd/cli/kubectl-kyverno/policy/annotations"
```
Please remove commented lines...

JimBugwadia commented on pkg/utils/report/results.go:
```diff
@@ -86,72 +87,69 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
-			results = append(results, result)
+func ToPolicyReportResult(policyType engineapi.PolicyType, policyName string, ruleResult engineapi.RuleResponse, annotations map[string]string, resource *corev1.ObjectReference) policyreportv1alpha2.PolicyReportResult {
+	result := policyreportv1alpha2.PolicyReportResult{
+		Properties: map[string]string{},
```
lets move this to the end of the list of attributes...line 104

JimBugwadia reviewed (changes requested):
Looks good, a few minor comments on format and commented lines! 

Can you please also check if if all tests are working?

JimBugwadia commented:
Hi @coder12git - the code changes look fine, but the tests are failing here:

https://github.com/kyverno/kyverno/actions/runs/8105309068/job/22155994071#step:5:1896

Can you please take a look and fix?

```
=== RUN   TestComputePolicyReportResult/skip
    report_test.go:298: ComputePolicyReportResult() = {kyverno pod-requirements xxx [{      }] nil test skip true map[] {0 0} Pod Security Standards (Restricted) medium}, want {kyverno pod-requirements xxx [{      }] nil test skip true map[] {0 0} Pod Security Standards (Restricted) medium}
=== RUN   TestComputePolicyReportResult/pass
    report_test.go:298: ComputePolicyReportResult() = {kyverno pod-requirements xxx [{      }] nil test pass true map[] {0 0} Pod Security Standards (Restricted) medium}, want {kyverno pod-requirements xxx [{      }] nil test pass true map[] {0 0} Pod Security Standards (Restricted) medium}
=== RUN   TestComputePolicyReportResult/fail
    report_test.go:298: ComputePolicyReportResult() = {kyverno pod-requirements xxx [{      }] nil test fail true map[] {0 0} Pod Security Standards (Restricted) medium}, want {kyverno pod-requirements xxx [{      }] nil test fail true map[] {0 0} Pod Security Standards (Restricted) medium}

```

coder12git commented:
> === RUN   TestComputePolicyReportResult/skip
>     report_test.go:298: ComputePolicyReportResult() = {kyverno pod-requirements xxx [{      }] nil test skip true map[] {0 0} Pod Security Standards (Restricted) medium}, want {kyverno pod-requirements xxx [{      }] nil test skip true map[] {0 0} Pod Security Standards (Restricted) medium}
> === RUN   TestComputePolicyReportResult/pass
>     report_test.go:298: ComputePolicyReportResult() = {kyverno pod-requirements xxx [{      }] nil test pass true map[] {0 0} Pod Security Standards (Restricted) medium}, want {kyverno pod-requirements xxx [{      }] nil test pass true map[] {0 0} Pod Security Standards (Restricted) medium}
> === RUN   TestComputePolicyReportResult/fail
>     report_test.go:298: ComputePolicyReportResult() = {kyverno pod-requirements xxx [{      }] nil test fail true map[] {0 0} Pod Security Standards (Restricted) medium}, want {kyverno pod-requirements xxx [{      }] nil test fail true map[] {0 0} Pod Security Standards (Restricted) medium}

Sure, working on it

JimBugwadia commented:
The issue is caused by the empty map declaration at:

https://github.com/kyverno/kyverno/pull/9785/files#diff-aaaccc381801aa74e58e1e5df6113de7864414655ddfc5b2cf0aa3d29f32573cR103

Lets remove this and add a function that safely adds a property:

```go
func addProperty(k, v string, result policyreportv1alpha2.PolicyReportResult) {
	if result.Properties == nil {
		result.Properties = map[string]string{}
	}

	result.Properties[k] = v
}

```

We can then reuse this when we need to add a property:

https://github.com/kyverno/kyverno/pull/9785/files#diff-aaaccc381801aa74e58e1e5df6113de7864414655ddfc5b2cf0aa3d29f32573cR114

https://github.com/kyverno/kyverno/pull/9785/files#diff-aaaccc381801aa74e58e1e5df6113de7864414655ddfc5b2cf0aa3d29f32573cR126

coder12git commented:
> The issue is caused by the empty map declaration at:
> 
> https://github.com/kyverno/kyverno/pull/9785/files#diff-aaaccc381801aa74e58e1e5df6113de7864414655ddfc5b2cf0aa3d29f32573cR103
> 
> Lets remove this and add a function that safely adds a property:
> 
> ```go
> func addProperty(k, v string, result policyreportv1alpha2.PolicyReportResult) {
> 	if result.Properties == nil {
> 		result.Properties = map[string]string{}
> 	}
> 
> 	result.Properties[k] = v
> }
> ```
> 
> We can then reuse this when we need to add a property:
> 
> https://github.com/kyverno/kyverno/pull/9785/files#diff-aaaccc381801aa74e58e1e5df6113de7864414655ddfc5b2cf0aa3d29f32573cR114
> 
> https://github.com/kyverno/kyverno/pull/9785/files#diff-aaaccc381801aa74e58e1e5df6113de7864414655ddfc5b2cf0aa3d29f32573cR126

Sure, btw we need a pointer reference here for result while calling this in a function

coder12git commented on cmd/cli/kubectl-kyverno/report/report_test.go:
```diff
@@ -268,6 +268,27 @@ func TestComputePolicyReportResult(t *testing.T) {
 			Category:  "Pod Security Standards (Restricted)",
 			Severity:  policyreportv1alpha2.SeverityMedium,
 		},
+	}, {
+		name:           "fail",
+		auditWarn:      false,
+		engineResponse: engineapi.NewEngineResponse(unstructured.Unstructured{}, engineapi.NewKyvernoPolicy(policy), nil),
+		ruleResponse:   *engineapi.RuleWarn("xxx", engineapi.Mutation, "test"),
+		want: policyreportv1alpha2.PolicyReportResult{
+			Source: "kyverno",
+			Policy: "psa",
+			Properties: map[string]string{
+				"controls": "allowPrivilegeEscalation, capabilities_restricted, runAsNonRoot, seccompProfile_restricted",
+				"standard": "restricted",
+				"version":  "xxx",
+			},
+			Rule:      "xxx",
+			Result:    policyreportv1alpha2.StatusError,
+			Resources: []corev1.ObjectReference{{}},
+			Message:   "test",
+			Scored:    true,
+			Category:  "Pod Security Standards (Restricted)",
+			Severity:  policyreportv1alpha2.SeverityMedium,
+		},
```
I have added this test, but even if I remove it, still the tests are failing. I checked locally to via `make test-unit` command.

JimBugwadia replied:
You can check the errors in the run logs:

https://github.com/kyverno/kyverno/actions/runs/8107195987/job/22158313096?pr=9785#step:5:1888

If you run the test locally in a debugger, you should be able to see why the result does not match the expected result.

coder12git replied:
Yep

coder12git commented:
@JimBugwadia tests are passed now, please have a look, also i have deleted an library in report_test.yaml as it was imported two times and for warn test there were some minute error that was causing test failure. For PSS report, i have added a seperate policy in test_data , and have created a new function for testing the PSS report.

JimBugwadia commented:
@coder12git - looks like we now have some Chainsaw tests failing:

https://github.com/kyverno/kyverno/actions/runs/8112917352/job/22191989700?pr=9785#step:10:72

Can you please take a look?

coder12git commented:
> @coder12git - looks like we now have some Chainsaw tests failing:
> 
> https://github.com/kyverno/kyverno/actions/runs/8112917352/job/22191989700?pr=9785#step:10:72
> 
> Can you please take a look?

Not sure, about this error
```
   l.go:53: | 23:03:48 | validating-admission-policy-fail | step-03  | ASSERT    | ERROR | wgpolicyk8s.io/v1alpha2/PolicyReport @ chainsaw-tidy-buzzard/*
        === ERROR
        no actual resource found
```

coder12git commented:
@JimBugwadia is there any way to pass these chainsaw tests, I'm not able to find any errors  from logs.

JimBugwadia commented:
Hi @coder12git - to troubleshoot, you can manually try and run the same test locally. Like the logs, local execution shows that the test fails as the report is not generated.

I verified that this test passes with the prior version, but does not pass with the updated version.  I also verified that the report is generated if we run the CLI.

So, it seems like there is something specific to the ValidatingAdmissionPolicy report generation during admission controls or background scans.

I was not able to troubleshoot further but have requested @MariamFahmy98 to take a look at the changes and see if she spots any issues.

If there is nothing obvious, the next step would be to run this locally in a debugger and figure out why the report does not get generated.

MariamFahmy98 commented on pkg/utils/report/results.go:
```diff
@@ -86,72 +87,76 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
+
+func EngineResponseToReportResults(response engineapi.EngineResponse) []policyreportv1alpha2.PolicyReportResult {
+	pol := response.Policy()
+	policyName, _ := cache.MetaNamespaceKeyFunc(pol.AsKyvernoPolicy())
```
This is the issue. You have to check the policy type and according to it, get either the Kyverno policy name or the VAP name because now, you always get the Kyverno policy name. Moreover, the VAP name can be extracted from the ruleResponse, there is no need to get it from the cache.

JimBugwadia replied:
Thanks @MariamFahmy98! 

@coder12git - can you add me as a collaborator to your repo, so I can push changes?

JimBugwadia replied:
@MariamFahmy98 - would a `pol.GetName()` not work here for both Kyverno policies and ValidatingAdmissionPolicy types?

coder12git replied:
> @coder12git - can you add me as a collaborator to your repo, so I can push changes?

@JimBugwadia i have added you as a collaborator.

MariamFahmy98 replied:
> @MariamFahmy98 - would a `pol.GetName()` not work here for both Kyverno policies and ValidatingAdmissionPolicy types?

It will work.

MariamFahmy98 commented on pkg/utils/report/results.go:
```diff
@@ -86,72 +87,76 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
+	result := policyreportv1alpha2.PolicyReportResult{
+		Source:  kyverno.ValueKyvernoApp,
+		Policy:  policyName,
+		Rule:    ruleResult.Name(),
```
If the policy type is ValidatingAdmissionPolicy, the `rule` field should be unset.

JimBugwadia replied:
Will `ruleResult.Name()` not be an empty string then?

JimBugwadia replied:
I tried it and this seems to work:

```go
func EngineResponseToReportResults(response engineapi.EngineResponse) []policyreportv1alpha2.PolicyReportResult {
	policy := response.Policy()
	policyType := policy.GetType()
	policyName := policy.GetName()
	annotations := policy.GetAnnotations()

	var results []policyreportv1alpha2.PolicyReportResult
	for _, ruleResult := range response.PolicyResponse.Rules {
		result := ToPolicyReportResult(policyType, policyName, ruleResult, annotations, nil)
		results = append(results, result)
	}

	return results
}
```

JimBugwadia replied:
After this change, I was able to locally run the Chainsaw tests successfully:

```sh
chainsaw test --test-dir ./validating-admission-policy-reports --config .chainsaw.yaml
```

```sh
Version: 0.1.7
Loading config (.chainsaw.yaml)...
- Using test file: chainsaw-test.yaml
- TestDirs [./validating-admission-policy-reports]
- SkipDelete false
- FailFast true
- ReportFormat ''
- ReportName 'chainsaw-report'
- Namespace ''
- FullName true
- IncludeTestRegex ''
- ExcludeTestRegex '_.+'
- ApplyTimeout 5s
- AssertTimeout 1m30s
- CleanupTimeout 30s
- DeleteTimeout 15s
- ErrorTimeout 1m30s
- ExecTimeout 5s
- Parallel 1
- ForceTerminationGracePeriod 5s
- DelayBeforeCleanup 3s
- NoCluster false
Loading tests...
- validating-admission-policy-fail (validating-admission-policy-reports/background/validating-admission-policy-fail)
- validating-admission-policy-pass (validating-admission-policy-reports/background/validating-admission-policy-pass)
- with-validating-admission-policy-binding-fail (validating-admission-policy-reports/background/with-validating-admission-policy-binding-fail)
- with-validating-admission-policy-binding-no-reports (validating-admission-policy-reports/background/with-validating-admission-policy-binding-no-reports)
- with-validating-admission-policy-binding-pass (validating-admission-policy-reports/background/with-validating-admission-policy-binding-pass)
- events (validating-admission-policy-reports/events)
Loading values...
Running tests...
=== RUN   chainsaw
=== PAUSE chainsaw
=== CONT  chainsaw
=== RUN   chainsaw/validating-admission-policy-reports/background/validating-admission-policy-fail[validating-admission-policy-fail]
=== PAUSE chainsaw/validating-admission-policy-reports/background/validating-admission-policy-fail[validating-admission-policy-fail]
=== RUN   chainsaw/validating-admission-policy-reports/background/validating-admission-policy-pass[validating-admission-policy-pass]
=== PAUSE chainsaw/validating-admission-policy-reports/background/validating-admission-policy-pass[validating-admission-policy-pass]
=== RUN   chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-fail[with-validating-admission-policy-binding-fail]
=== PAUSE chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-fail[with-validating-admission-policy-binding-fail]
=== RUN   chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-no-reports[with-validating-admission-policy-binding-no-reports]
=== PAUSE chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-no-reports[with-validating-admission-policy-binding-no-reports]
=== RUN   chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-pass[with-validating-admission-policy-binding-pass]
=== PAUSE chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-pass[with-validating-admission-policy-binding-pass]
=== RUN   chainsaw/validating-admission-policy-reports/events[events]
=== PAUSE chainsaw/validating-admission-policy-reports/events[events]
=== CONT  chainsaw/validating-admission-policy-reports/background/validating-admission-policy-fail[validating-admission-policy-fail]
    | 10:54:19 | validating-admission-policy-fail | @setup   | CREATE    | OK    | v1/Namespace @ chainsaw-choice-snipe
    | 10:54:19 | validating-admission-policy-fail | step-01  | TRY       | RUN   |
    | 10:54:19 | validating-admission-policy-fail | step-01  | APPLY     | RUN   | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | CREATE    | OK    | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | APPLY     | DONE  | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | ASSERT    | RUN   | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | ASSERT    | DONE  | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | TRY       | DONE  |
    | 10:54:19 | validating-admission-policy-fail | step-02  | TRY       | RUN   |
    | 10:54:19 | validating-admission-policy-fail | step-02  | APPLY     | RUN   | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | CREATE    | OK    | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | APPLY     | DONE  | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | ASSERT    | RUN   | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | ASSERT    | DONE  | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | TRY       | DONE  |
    | 10:54:19 | validating-admission-policy-fail | step-03  | TRY       | RUN   |
    | 10:54:19 | validating-admission-policy-fail | step-03  | ASSERT    | RUN   | wgpolicyk8s.io/v1alpha2/PolicyReport @ chainsaw-choice-snipe/*
    | 10:54:29 | validating-admission-policy-fail | step-03  | ASSERT    | DONE  | wgpolicyk8s.io/v1alpha2/PolicyReport @ chainsaw-choice-snipe/*
    | 10:54:29 | validating-admission-policy-fail | step-03  | TRY       | DONE  |
    | 10:54:32 | validating-admission-policy-fail | @cleanup | DELETE    | RUN   | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:32 | validating-admission-policy-fail | @cleanup | DELETE    | OK    | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:32 | validating-admission-policy-fail | @cleanup | DELETE    | DONE  | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    
    ...

Tests Summary...
- Passed  tests 6
- Failed  tests 0
- Skipped tests 0
Done.
```

MariamFahmy98 commented on pkg/utils/report/results.go:
```diff
@@ -86,72 +87,76 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
+		Timestamp: metav1.Timestamp{
+			Seconds: time.Now().Unix(),
+		},
+		Category: annotations[kyverno.AnnotationPolicyCategory],
+		Severity: SeverityFromString(annotations[kyverno.AnnotationPolicySeverity]),
```
These annotations are specific to Kyverno policies. They should be unset for VAPs.

JimBugwadia replied:
If they are not set, does it matter?

MariamFahmy98 replied:
I am uncertain about the purpose of `category `and `severity` in generated reports and what they are used for. However, we can add them to VAPs using either Kyverno annotations or specific VAP annotations.

MariamFahmy98 commented on pkg/utils/report/results.go:
```diff
@@ -86,72 +87,76 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
+		Rule:    ruleResult.Name(),
+		Message: ruleResult.Message(),
+		Result:  toPolicyResult(ruleResult.Status()),
+		Scored:  annotations[kyverno.AnnotationPolicyScored] != "false",
```
This annotation is specific to Kyverno policies. It should be unset for VAPs.

JimBugwadia replied:
If the annotation is not present, it will default to true, right? Would that be a problem.

MariamFahmy98 commented on cmd/cli/kubectl-kyverno/report/report.go:
```diff
@@ -13,47 +11,26 @@ import (
 
 func ComputePolicyReportResult(auditWarn bool, engineResponse engineapi.EngineResponse, ruleResponse engineapi.RuleResponse) policyreportv1alpha2.PolicyReportResult {
```
Since `engineResponse` already has the `ruleResponse`, why do we need to pass both?
If there is only the `engineResponse` parameter, we can use `EngineResponseToReportResults()` as it is without altering its behavior.

MariamFahmy98 replied:
So `ComputePolicyReportResult()` takes both `engineResponse` and `ruleResponse`. I wonder why does it take both? As `engineResponse` already has the `policyResponse` which in return has the `ruleResponse`. If we could make this function takes only the `engineResponse`, then we can safely call `EngineResponseToReportResults()` without making any code changes.

This is the current [EngineResponseToReportResults()](https://github.com/kyverno/kyverno/blob/main/pkg/utils/report/results.go#L89) used by reports controller.

coder12git commented:
THanks @MariamFahmy98 I will do these changes ASAP.

coder12git commented:
Now most of the chainsaw tests are passed (except policy library)- https://github.com/kyverno/kyverno/actions/runs/8156925619/job/22296784852?pr=9785

JimBugwadia commented:
@coder12git - looks like the only test still failing (across all versions) is:

  https://github.com/kyverno/kyverno/actions/runs/8156925619/job/22304258617?pr=9785#step:14:3752

This corresponds to the test case here:

  https://github.com/kyverno/policies/tree/main/other/record-creation-details/.chainsaw-test

Interestingly there were some changes made a few days ago.

I ran it locally and see this error:

```sh
=== RUN   chainsaw/other/record-creation-details/.chainsaw-test[record-creation-details]
=== PAUSE chainsaw/other/record-creation-details/.chainsaw-test[record-creation-details]
=== CONT  chainsaw/other/record-creation-details/.chainsaw-test[record-creation-details]
=== NAME  chainsaw
    | 23:16:37 | chainsaw | @main | INTERNAL  | ERROR |
        === ERROR
        <nil>: Internal error: invalid type for: <nil>, expected: []functions.JpType{"string"}
--- FAIL: chainsaw (0.00s)
    --- FAIL: chainsaw/other/record-creation-details/.chainsaw-test[record-creation-details] (0.00s)
FAIL
Tests Summary...
- Passed  tests 0
- Failed  tests 1
- Skipped tests 0
Done with failures.
Error: some tests failed
```

@eddycharly  - any idea what could be causing this error?

JimBugwadia commented:
Looks like this is the issue - the test command was changed:

https://github.com/kyverno/policies/pull/918/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R114

JimBugwadia commented on pkg/utils/report/results.go:
```diff
@@ -156,7 +156,7 @@ func ToPolicyReportResult(policyType engineapi.PolicyType, policyName string, ru
 func EngineResponseToReportResults(response engineapi.EngineResponse) []policyreportv1alpha2.PolicyReportResult {
 	pol := response.Policy()
-	policyName := pol.GetName()
+	policyName, _ := cache.MetaNamespaceKeyFunc(pol.AsKyvernoPolicy())
```
@coder12git - any reason why you reverted this?

coder12git replied:
yes, because i have added if and else statement inside ToPolicyReport function, so if it will be VPA, in that case policy name will be extracted from rule response.

coder12git replied:
```
else {
		result = policyreportv1alpha2.PolicyReportResult{
			Source:  "ValidatingAdmissionPolicy",
			Policy:  ruleResult.Name(),
			Message: ruleResult.Message(),
			Result:  toPolicyResult(ruleResult.Status()),
			Timestamp: metav1.Timestamp{
				Seconds: time.Now().Unix(),
			},
		}
		if ruleResult.ValidatingAdmissionPolicyBinding() != nil {
			addProperty("binding", ruleResult.ValidatingAdmissionPolicyBinding().Name, &result)
		}
	}
```

JimBugwadia replied:
Its best to replace type based if-else statements, by delegating to the type itself.

I will take a look and update.

realshuting commented:
Which milestone are we targeting?

coder12git commented on pkg/utils/report/results.go:
```diff
@@ -86,72 +87,76 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
+		}
+	}
+	if policyType == engineapi.ValidatingAdmissionPolicyType {
+		result.Source = "ValidatingAdmissionPolicy"
```
@JimBugwadia category and severity field are added here, which we don't need in case of VAP, I think that's the reason why one of the chainsaw test is still failing.

JimBugwadia replied:
@coder12git - will take a look. 

We do want those, if they are declared in the ValidatingAdmissionPolicy, as annotations.

coder12git replied:
That's the reason why i added if and else statement, to remove those fields in case of VAP.
